velox/exec/Task.cpp

```diff
@@ -1644,4 +1653,90 @@ void Task::TaskCompletionNotifier::notify() {
     active_ = false;
   }
 }
+
+/// Example Error Message generated:
+/// Exceeded memory cap of 12.00MB when requesting 1.00MB. Task total: 8.00MB
+/// Peak: 12.00MB. Top 3 Operators (by aggregate usage across all instances):
+/// Aggregation_#1_x6: 168.00KB Peak: 1.00MB, TableScan_#0_x6: 51.46KB
+/// Peak: 1.00MB, CallbackSink_#2_x6: 0B Peak: 0B. Failed Operator:
+/// Aggregation_#1: 0B
+std::string Task::getErrorMsgOnMemCapExceeded(
+    memory::MemoryUsageTracker& tracker) {
+  std::stringstream out;
+  out << "Task total: " << succinctBytes(tracker.getCurrentTotalBytes())
+      << " Peak: " << succinctBytes(tracker.getPeakTotalBytes()) << ".";
+  // Aggregate relevant metrics for each operator across all drivers.
+  struct MemoryUsageStats {
+    int operatorId;
+    std::string operatorType;
+    int64_t cumulativeTotalBytes;
+    int64_t peakTotalBytes;
+    int numInstances;
+  };
+  std::unordered_map<int, MemoryUsageStats> operatorStats;
+  {
+    std::lock_guard<std::mutex> l(mutex_);
+    for (auto& driver : drivers_) {
+      if (!driver) {
+        continue;
+      }
+      auto operatorsInCurrDriver = driver->operators();
+      for (auto op : operatorsInCurrDriver) {
+        auto itr = operatorStats.find(op->stats().operatorId);
+        if (itr == operatorStats.end()) {
+          itr = operatorStats
+                    .insert(
+                        {op->stats().operatorId,
+                         {op->stats().operatorId,
```

Comment:
nit: It might be easier to read this code if we add MemoryUsageStats constructor that takes just operator ID and type or just OperatorStats.